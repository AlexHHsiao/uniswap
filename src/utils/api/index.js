const BASE_URL = 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v3';
// function to fetch all tokens
export const getTokens = () => {
    const query = `
        {
            tokens(orderBy: totalValueLockedUSD, orderDirection: desc) {
                id
                symbol
                name
                totalValueLockedUSD
                tokenDayData(first: 3, orderBy: date, orderDirection: desc) {
                    priceUSD
                    volumeUSD
                    date
                }
            }
        }
    `;

    return fetch(`${BASE_URL}`, {
        method: 'POST',
        body: JSON.stringify({query})
    });
};

// function to fetch all Pools
export const getPools = () => {
    const query = `
        {
            pools(orderBy: totalValueLockedUSD, orderDirection: desc) {
                id
                totalValueLockedUSD
                poolDayData(first:1, orderBy:date, orderDirection: desc) {
                    volumeUSD
                }
                token0 {
                    id
                    symbol
                    name
                }
                token1 {
                    id
                    symbol
                    name
                }
            }
        }
    `;

    return fetch(`${BASE_URL}`, {
        method: 'POST',
        body: JSON.stringify({query})
    });
};

// function to fetch all transactions
export const getTransactions = () => {
    const query = `
        {
            transactions(orderBy: timestamp, orderDirection: desc) {
                id
                timestamp
                burns(first:1, orderBy:timestamp, orderDirection: desc) {
                    origin
                    amountUSD
                    amount0
                    amount1
                    token0 {
                        id
                        symbol
                        name
                    }
                    token1 {
                        id
                        symbol
                        name
                    }
                }
                swaps(first:1, orderBy:timestamp, orderDirection: desc) {
                    recipient
                    amountUSD
                    amount0
                    amount1
                    token0 {
                        id
                        symbol
                        name
                    }
                    token1 {
                        id
                        symbol
                        name
                    }
                }
                mints(first:1, orderBy:timestamp, orderDirection: desc) {
                    origin
                    amountUSD
                    amount0
                    amount1
                    token0 {
                        id
                        symbol
                        name
                    }
                    token1 {
                        id
                        symbol
                        name
                    }
                }
            }
        }
    `;

    return fetch(`${BASE_URL}`, {
        method: 'POST',
        body: JSON.stringify({query})
    });
};