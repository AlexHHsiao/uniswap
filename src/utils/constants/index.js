export const UNISWAP_URL = 'https://info.uniswap.org/';
export const ETHERSCAN_URL = 'https://etherscan.io/';