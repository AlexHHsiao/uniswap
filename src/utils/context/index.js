import {createContext} from "react";

// app context will be used to store spinner and message data
export const AppContext = createContext(null);