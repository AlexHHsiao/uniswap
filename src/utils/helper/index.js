import {ETHERSCAN_URL} from "../constants";

// helper method to parse token data into table format
export const parseTokenHelper = (tokens) =>
    tokens.map(({id, name, symbol, totalValueLockedUSD, tokenDayData}) => {
        let price = '0';
        let priceChange = 0;

        if (tokenDayData.length > 0) {
            const {priceUSD} = tokenDayData[0];
            price = priceUSD;
        }

        if (tokenDayData.length > 1 && tokenDayData[tokenDayData.length - 1].priceUSD !== '0') {
            const {priceUSD} = tokenDayData[tokenDayData.length - 1];
            priceChange = (price - priceUSD) / priceUSD * 100;
        }

        return {
            key: id,
            priceChange,
            price: parseInt(price),
            TVL: parseInt(totalValueLockedUSD),
            name: {name, symbol}
        }
    });

// helper method to parse pool data into table format
export const parsePoolHelper = (pools) =>
    pools.map(({
                   id,
                   totalValueLockedUSD,
                   token0: {symbol: symbol1},
                   token1: {symbol: symbol2},
                   poolDayData: [{volumeUSD}]
               }) => ({
        key: id,
        volumeUSD: parseInt(volumeUSD),
        TVL: parseInt(totalValueLockedUSD),
        pool: `${symbol1}/${symbol2}`
    }));

// helper method to parse transaction data into table format
export const parseTransactionHelper = (transactions) =>
    transactions.map(({id, timestamp, burns, swaps, mints}) => {
        let action;
        let tokenAmount0;
        let tokenAmount1;
        let totalValue;
        let account;

        if (burns.length > 0) {
            const [{origin, amountUSD, amount0, amount1, token0: {symbol: symbol0}, token1: {symbol: symbol1}}] = burns;
            action = {id, name: `Remove ${symbol1} and ${symbol0}`};
            tokenAmount0 = {symbol: symbol0, value: Math.abs(parseInt(amount0))};
            tokenAmount1 = {symbol: symbol1, value: Math.abs(parseInt(amount1))};
            account = origin;
            totalValue = parseInt(amountUSD);
        } else if (swaps.length > 0) {
            const [{
                recipient,
                amountUSD,
                amount0,
                amount1,
                token0: {symbol: symbol0},
                token1: {symbol: symbol1}
            }] = swaps;
            action = {id, name: `Swap ${symbol1} for ${symbol0}`};
            tokenAmount0 = {symbol: symbol0, value: Math.abs(parseInt(amount0))};
            tokenAmount1 = {symbol: symbol1, value: Math.abs(parseInt(amount1))};
            account = recipient;
            totalValue = parseInt(amountUSD);
        } else {
            const [{origin, amountUSD, amount0, amount1, token0: {symbol: symbol0}, token1: {symbol: symbol1}}] = mints;
            action = {id, name: `Add ${symbol1} and ${symbol0}`};
            tokenAmount0 = {symbol: symbol0, value: Math.abs(parseInt(amount0))};
            tokenAmount1 = {symbol: symbol1, value: Math.abs(parseInt(amount1))};
            account = origin;
            totalValue = parseInt(amountUSD);
        }

        return {
            key: id,
            action,
            timestamp,
            tokenAmount0,
            tokenAmount1,
            account,
            totalValue
        }
    });

export const etherscanUserLocation = (id) => `${ETHERSCAN_URL}address/${id}`;

export const etherscanTxLocation = (id) => `${ETHERSCAN_URL}tx/${id}`;

// helper method format time difference from now
export const timeCal = (timestamp) => {
    const date = new Date(timestamp * 86400);
    const now = new Date();
    const diffInMs = now.getTime() - date.getTime();

    // if the time difference is less than 1 minute
    if (diffInMs < 60 * 1000) {
        return 'Recently';
    }
    // if the time difference is less than 1 hour
    else if (diffInMs < 60 * 60 * 1000) {
        const diffInMin = Math.round(diffInMs / (1000 * 60));
        return `${diffInMin} mins ago`;
    }
    // if the time difference is less than 1 day
    else if (diffInMs < 24 * 60 * 60 * 1000) {
        const diffInHrs = Math.round(diffInMs / (1000 * 60 * 60));
        return `${diffInHrs} hours ago`;
    }
    // if the time difference is more than 1 day
    else {
        const diffInDays = Math.round(diffInMs / (1000 * 60 * 60 * 24));
        return `${diffInDays} days ago`;
    }
};

// helper method to format big number with unit
export const numUnitHelper = (number) => {
    let num;

    if (number >= 1000000000) {
        num = `${(number / 1000000000).toFixed(2)}b`;
    } else if (number >= 1000000) {
        num = `${(number / 1000000).toFixed(2)}m`;
    } else if (number >= 1000) {
        num = `${(number / 1000).toFixed(2)}k`;
    } else {
        num = number.toFixed(2);
    }

    return num;
};