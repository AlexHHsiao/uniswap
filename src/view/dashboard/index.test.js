import renderer from 'react-test-renderer';
import Dashboard from './index';
import {AppContext} from "../../utils/context";

const setSpinnerLoading = jest.fn();
const setMessageData = jest.fn();
const MyComponent = () =>
    <AppContext.Provider value={{setSpinnerLoading, setMessageData}}>
        <Dashboard/>
    </AppContext.Provider>

jest.mock("antd", () => {
    const lib = jest.requireActual("antd");

    return {
        ...lib,
        Tabs: () => <div></div>,
    };
});

describe("Dashboard Tests", () => {
    it('Matches DOM Snapshot', () => {
        const tree = renderer
            .create(<MyComponent/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});