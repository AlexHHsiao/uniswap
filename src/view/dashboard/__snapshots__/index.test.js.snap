// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Tests Matches DOM Snapshot 1`] = `<div />`;
